op-service/backoff: Refactor to use time.Duration instead of float/int.

Rcontre360 opened the pull request:
changed backoff/strategies.go and backoff/strategies_test.go, only types and a function changed

<!-- Contributions welcome! See https://github.com/ethereum-optimism/.github/blob/master/CONTRIBUTING.md -->

**Description**

No features added. Just refactored backoff package to allow only types from time.Duration on time related variables. 

**Tests**

Only modified a single test to use time.Duration instead of int.

**Metadata**

- Fixes #[6730](https://github.com/ethereum-optimism/optimism/issues/6730#issuecomment-1678950667)

protolambda commented on op-service/backoff/strategies.go:
```diff
@@ -43,8 +43,8 @@ func (e *ExponentialStrategy) Duration(attempt int) time.Duration {
-		Max:       10000,
-		MaxJitter: 250,
+		Max:       time.Duration(10000),
+		MaxJitter: time.Duration(250),
```
```suggestion
		Max:       time.Duration(10000 * time.Millisecond),
		MaxJitter: time.Duration(250 * time.Millisecond),
```

protolambda commented on op-service/backoff/strategies.go:
```diff
@@ -17,23 +17,23 @@ type Strategy interface {
 // function is min(e.Min + (2^attempt * 1000) + randBetween(0, e.MaxJitter), e.Max)
 type ExponentialStrategy struct {
 	// Min is the minimum amount of time to wait between attempts in ms.
```
```suggestion
	// Min is the minimum amount of time to wait between attempts.
```

protolambda commented on op-service/backoff/strategies.go:
```diff
@@ -17,23 +17,23 @@ type Strategy interface {
-	Min float64
+	Min time.Duration
 
 	// Max is the maximum amount of time to wait between attempts in ms.
```
```suggestion
	// Max is the maximum amount of time to wait between attempts.
```

protolambda commented on op-service/backoff/strategies.go:
```diff
@@ -17,23 +17,23 @@ type Strategy interface {
-	Max float64
+	Max time.Duration
 
 	// MaxJitter is the maximum amount of random jitter to insert between
 	// attempts in ms.
```
```suggestion
	// MaxJitter is the maximum amount of random jitter to insert between attempts.
```

protolambda commented on op-service/backoff/strategies.go:
```diff
@@ -17,23 +17,23 @@ type Strategy interface {
+	var jitter time.Duration
 	if e.MaxJitter > 0 {
-		jitter = rand.Intn(e.MaxJitter)
+		jitter = time.Duration(rand.Int63n(e.MaxJitter.Milliseconds()))
 	}
-	dur := e.Min + (math.Pow(2, float64(attempt)) * 1000)
-	dur += float64(jitter)
+	dur := e.Min + time.Duration((math.Pow(2, float64(attempt)) * 1000))
+	dur += jitter
```
Also needs corrections. You can't just convert from int to time.Duration, it has to scale with the correct unit.

Rcontre360 replied:
Got it

protolambda commented on op-service/backoff/strategies_test.go:
```diff
@@ -14,8 +14,8 @@ func TestExponential(t *testing.T) {
 
-	durations := []int{4, 5, 7, 10, 10}
+	durations := []time.Duration{4, 5, 7, 10, 10}
 	for i, dur := range durations {
-		require.Equal(t, time.Millisecond*time.Duration(dur*1000), strategy.Duration(i))
+		require.Equal(t, time.Millisecond*dur*1000, strategy.Duration(i))
```
Would revert this, or change the values in the array to multiples of `time.Millisecond`

trianglesphere commented on op-service/backoff/strategies.go:
```diff
@@ -16,24 +16,23 @@ type Strategy interface {
+	var jitter time.Duration
 	if e.MaxJitter > 0 {
-		jitter = rand.Intn(e.MaxJitter)
+		jitter = time.Duration(rand.Int63n(e.MaxJitter.Milliseconds()))
```
`time.Duration` is in nanoseconds, when creating one, you always need to multiple by the unit.

Rcontre360 replied:
Yes, thanks! didnt know it when making the changes

Rcontre360 reviewed (commented):
There's only one place where ExponentialStrategy is used, it is /op-node/sources/sync_client.go. It is assumed in the expStrategy. Duration function that the fields in the expStrategy will be provided correctly (example: Max: time.Duration(1000 * time.Millisecond)).

protolambda commented on op-service/backoff/strategies.go:
```diff
@@ -29,15 +29,15 @@ type ExponentialStrategy struct {
 	}
 
-	return time.Millisecond * time.Duration(dur)
+	return time.Duration(dur)
```
```suggestion
	return dur
```
It's already a `time.Duration`

protolambda commented on op-service/backoff/strategies_test.go:
```diff
@@ -9,13 +9,14 @@ import (
 	durations := []time.Duration{4, 5, 7, 10, 10}
 	for i, dur := range durations {
-		require.Equal(t, time.Millisecond*dur*1000, strategy.Duration(i))
+		msDuration := dur * time.Millisecond * 1000
+		require.Equal(t, msDuration, strategy.Duration(i))
```
```suggestion
		require.Equal(t, dur*time.Second, strategy.Duration(i))
```

protolambda reviewed (commented):
Two more nits, then done

protolambda commented:
And `ExponentialStrategy` in `sync_client.go` needs the `* time.Millisecond` to have the proper units

Rcontre360 commented:
Yes @protolambda I already changed op-node/sources/sync_client.go to have the correct units :+1:

tynes commented:
@Rcontre360 if you rebase, CI should be passing again

Rcontre360 commented:
> @Rcontre360 if you rebase, CI should be passing again

hmmm. Just rebased and they keep failing :thinking: . Is this happening to other PRs as well?

protolambda commented:
@Rcontre360 likely other PRs as well, we're looking into it

protolambda commented:
@Mergifyio squash title+body

protolambda commented:
@Mergifyio squash title+body

protolambda committed:
op-service/backoff: Refactor to use time.Duration instead of float/int. (#6823)

changed backoff/strategies.go and backoff/strategies_test.go, only types and a function changed

**Description**

No features added. Just refactored backoff package to allow only types from time.Duration on time related variables.

**Tests**

Only modified a single test to use time.Duration instead of int.

**Metadata**

- Fixes #[[Link to Issue](https://github.com/ethereum-optimism/optimism/issues/6730#issuecomment-1678950667)]

Co-authored-by: Rcontre360 <rcontreraspimentel@gmail.com>

protolambda commented:
Ugh, mergify doesn't preserve the co-author info, let me fix that

protolambda commented:
fixed committer info, and rebased on develop to fix the flake

mergify[bot] committed:
Merge branch 'develop' into fix/backoff-package

mergify[bot] committed:
Merge branch 'develop' into fix/backoff-package
